File: python/src/cccc_sdk/client_0434_ops.py
from __future__ import annotations

from typing import Any, Dict, Iterable, Optional

from .client_0430_shared import _compact


class CCCC0434OpsMixin:
    """Operations added to the daemon IPC contract for CCCC 0.4.34."""

    def terminal_snapshot(
        self,
        *,
        group_id: str,
        actor_id: str,
        by: str = "user",
        limit_bytes: Optional[int] = None,
    ) -> Dict[str, Any]:
        return self.call(
            "terminal_snapshot",
            _compact(
                {
                    "group_id": str(group_id),
                    "actor_id": str(actor_id),
                    "by": str(by),
                    "limit_bytes": int(limit_bytes) if limit_bytes is not None else None,
                }
            ),
        )

    def web_model_delivery_preferences_get(
        self,
        *,
        group_id: str,
        actor_id: str,
    ) -> Dict[str, Any]:
        return self.call(
            "web_model_delivery_preferences_get",
            {"group_id": str(group_id), "actor_id": str(actor_id)},
        )

    def web_model_delivery_preferences_update(
        self,
        *,
        group_id: str,
        actor_id: str,
        mode: str,
        by: str = "user",
    ) -> Dict[str, Any]:
        normalized_mode = str(mode).strip()
        if normalized_mode not in {"standard", "image_compat"}:
            raise ValueError("mode must be 'standard' or 'image_compat'")
        return self.call(
            "web_model_delivery_preferences_update",
            {
                "group_id": str(group_id),
                "actor_id": str(actor_id),
                "mode": normalized_mode,
                "by": str(by),
            },
        )

    def web_model_runtime_recover_turn(
        self,
        *,
        group_id: str,
        actor_id: str,
        event_ids: Iterable[str],
    ) -> Dict[str, Any]:
        if isinstance(event_ids, (str, bytes)):
            raise ValueError("event_ids must be an iterable of event id strings")
        canonical_event_ids = [str(event_id).strip() for event_id in event_ids]
        if not canonical_event_ids or any(not event_id for event_id in canonical_event_ids):
            raise ValueError("event_ids must contain at least one non-empty event id")
        return self.call(
            "web_model_runtime_recover_turn",
            {
                "group_id": str(group_id),
                "actor_id": str(actor_id),
                "event_ids": canonical_event_ids,
            },
        )

File: ts/src/client_0434_ops.ts
import {
  compactRecord,
  type CCCC0430Client,
} from './client_0430_shared.js';
import type {
  TerminalSnapshotOptions,
  TerminalSnapshotResult,
  WebModelDeliveryPreferencesGetOptions,
  WebModelDeliveryPreferencesResult,
  WebModelDeliveryPreferencesUpdateOptions,
  WebModelRuntimeRecoverTurnOptions,
  WebModelRuntimeRecoverTurnResult,
} from './types.js';

export interface CCCC0434Ops {
  terminalSnapshot(options: TerminalSnapshotOptions): Promise<TerminalSnapshotResult>;
  webModelDeliveryPreferencesGet(
    options: WebModelDeliveryPreferencesGetOptions,
  ): Promise<WebModelDeliveryPreferencesResult>;
  webModelDeliveryPreferencesUpdate(
    options: WebModelDeliveryPreferencesUpdateOptions,
  ): Promise<WebModelDeliveryPreferencesResult>;
  webModelRuntimeRecoverTurn(
    options: WebModelRuntimeRecoverTurnOptions,
  ): Promise<WebModelRuntimeRecoverTurnResult>;
}

const ops: CCCC0434Ops & ThisType<CCCC0430Client> = {
  async terminalSnapshot(options) {
    const result = await this.call('terminal_snapshot', compactRecord({
      group_id: options.groupId,
      actor_id: options.actorId,
      by: options.by ?? 'user',
      limit_bytes: options.limitBytes,
    }));
    return result as unknown as TerminalSnapshotResult;
  },

  async webModelDeliveryPreferencesGet(options) {
    const result = await this.call('web_model_delivery_preferences_get', {
      group_id: options.groupId,
      actor_id: options.actorId,
    });
    return result as unknown as WebModelDeliveryPreferencesResult;
  },

  async webModelDeliveryPreferencesUpdate(options) {
    const result = await this.call('web_model_delivery_preferences_update', {
      group_id: options.groupId,
      actor_id: options.actorId,
      mode: options.mode,
      by: options.by ?? 'user',
    });
    return result as unknown as WebModelDeliveryPreferencesResult;
  },

  async webModelRuntimeRecoverTurn(options) {
    if (options.eventIds.length === 0 || options.eventIds.some((eventId) => eventId.trim().length === 0)) {
      throw new TypeError('eventIds must contain at least one non-empty event id');
    }
    const result = await this.call('web_model_runtime_recover_turn', {
      group_id: options.groupId,
      actor_id: options.actorId,
      event_ids: options.eventIds,
    });
    return result as unknown as WebModelRuntimeRecoverTurnResult;
  },
};

export function installCCCC0434Ops(proto: CCCC0430Client & Partial<CCCC0434Ops>): void {
  Object.assign(proto, ops);
}
